File: app/src/main/java/id/kuato/woahelper/main/SettingsButton.java
package id.kuato.woahelper.main;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.GradientDrawable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

import id.kuato.woahelper.R;
import id.kuato.woahelper.databinding.SettingsButtonBinding;

public class SettingsButton extends FrameLayout {
    private final SettingsButtonBinding layout;
    private final Context context;

    public SettingsButton(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        layout = SettingsButtonBinding.inflate(LayoutInflater.from(context), this, true);
        this.context = context;
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.SettingsButton, 0, 0);
        if (null == attrs) return;
        String text = a.getString(R.styleable.SettingsButton_text);
        layout.text.setText(text);
        setCornerRadius();
        setOnChangeListener((state) -> {
        });
    }

    void setOnChangeListener(OnChangeClickListener onChangeListener) {
        layout.switchButton.setOnCheckedChangeListener((v, state) -> {
            setCornerRadius();
            onChangeListener.onClick(state);
        });
    }

    public void setChecked(boolean state) {
        layout.switchButton.setChecked(state);
    }

    private int getColor() {
        return (layout.switchButton.isChecked()) ? 0xFF196500 : 0xFF870002;
    }

    private void setCornerRadius() {
        GradientDrawable background = new GradientDrawable();
        background.setColor(getColor());
        background.setCornerRadius(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 5.0f, context.getResources().getDisplayMetrics()));
        layout.getRoot().setBackground(background);
    }

    public interface OnChangeClickListener {
        void onClick(boolean state);
    }
}

File: app/src/main/java/com/woa/helper/main/SettingsButton.kt
package com.woa.helper.main

import android.content.Context
import android.graphics.drawable.GradientDrawable
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.widget.CompoundButton
import android.widget.FrameLayout
import com.woa.helper.R
import com.woa.helper.databinding.SettingsButtonBinding

class SettingsButton(private val context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {
    private val layout: SettingsButtonBinding = SettingsButtonBinding.inflate(LayoutInflater.from(context), this, true)

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.SettingsButton, 0, 0)
        if (null != attrs) {
            val text = a.getString(R.styleable.SettingsButton_text)
            layout.text.text = text
            setCornerRadius()
            setOnChangeListener { state: Boolean -> }
        }
    }

    fun setOnChangeListener(onChangeListener: OnChangeClickListener) {
        layout.switchButton.setOnCheckedChangeListener { v: CompoundButton?, state: Boolean ->
            setCornerRadius()
            onChangeListener.onClick(state)
        }
    }

    fun setChecked(state: Boolean) {
        layout.switchButton.isChecked = state
    }

    private val color: Int
        get() = if (layout.switchButton.isChecked) -0xe69b00 else -0x78fffe

    private fun setCornerRadius() {
        val background = GradientDrawable()
        background.setColor(this.color)
        background.cornerRadius = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 5.0f, context.resources.displayMetrics)
        layout.getRoot().background = background
    }

    fun interface OnChangeClickListener {
        fun onClick(state: Boolean)
    }
}
